cache-prefetch.go

```diff
@@ -0,0 +1,149 @@
+	}
+
+	if c.CacheTTLPollingCheckInterval == 0 {
+		c.CacheTTLPollingCheckInterval = time.Minute * 2
+	}
+	if c.RecordQueryHitsMin == 0 || c.RecordQueryHitsMin == 1 || c.RecordQueryHitsMin == -1 {
+		// Set to hit after one record hit
+		// fetch opportunistically
+		c.RecordQueryHitsMin = 1
+	}
+	if c.PrefetchSize == 0 {
+		c.PrefetchSize = 1000 // defaults to 1000 should be 1/4th of the cache size
+	}
+	go c.startCachePrefetchJobs()
+	return c
+}
+
+func (r *CachePrefetch) Resolve(q *dns.Msg, ci ClientInfo) (*dns.Msg, error) {
+
+	if len(q.Question) < 1 {
+		return nil, errors.New("no question in query")
+	}
+	// While multiple questions in one DNS message is part of the standard,
+	// it's not actually supported by servers. If we do get one of those,
+	// just pass it through and bypass caching.
+	if len(q.Question) > 1 {
+		return r.resolver.Resolve(q, ci)
+	}
+
+	// Get a response from upstream
+	a, err := r.resolver.Resolve(q.Copy(), ci)
+	if err != nil || a == nil {
+		return nil, err
+	}
+	//if rCode(a) == dns.RcodeToString[dns.RcodeSuccess] {
+		r.requestAddPrefetchJob(q)
+	//}
+
+	// Put the upstream response into the cache and return it. Need to store
+	// a copy since other elements might modify the response, like the replacer.
+	return a, nil
+}
+
+func (r *CachePrefetch) String() string {
+	return r.id
+}
+func (r *CachePrefetch) startCachePrefetchJobs() {
+	Log.WithFields(logrus.Fields{"id": r.id}).Trace("starting prefetching job")
+	for {
+		time.Sleep(r.CacheTTLPollingCheckInterval)
+		domainEntriesLength := len(r.metrics.items)
+		if domainEntriesLength > 0 {
+			for index, entry := range r.metrics.items {
+				Log.WithFields(logrus.Fields{"index": index, "total": domainEntriesLength}).Trace("prefetch")
+				r.startCachePrefetchJob(entry)
+			}
+		}
+	}
+}
+func (r *CachePrefetch) startCachePrefetchJob(item *CachePrefetchEntry) {
+	if (item == nil) || (item.msg == nil) || len(item.msg.Question) < 1 {
+		return
+	}
+
+	if item.prefetchState == PrefetchStateActive { // only prefetch if status is 1
+		qname := qName(item.msg)
+		qtype := qType(item.msg)
+
+		Log.WithFields(logrus.Fields{"qname": qname, "qtype": qtype}).Trace("prefetch request started")
+		var ci ClientInfo
+		a, err := r.resolver.Resolve(item.msg.Copy(), ci)
+
+		if err != nil || a == nil {
+			Log.WithFields(logrus.Fields{"err": err}).Trace("prefetch error")
+			r.mu.Lock()
+			r.metrics.addError(item.msg)
+			r.mu.Unlock()
+		} else {
+			if item.errorCount > 0 {
+				// reset error count after a successful request
+				r.mu.Lock()
+				r.metrics.resetError(item.msg)
+				r.mu.Unlock()
+				Log.WithFields(logrus.Fields{"qname": qname, "qtype": qtype}).Trace("query reset error count")
+			}
+			Log.WithFields(logrus.Fields{"qname": qname, "qtype": qtype}).Debug("query prefetched")
+		}
+	}
+}
+
+func (r *CachePrefetch) requestAddPrefetchJob(q *dns.Msg) {
+	if (q == nil) || len(q.Question) < 1 {
```

Comment:
No need to check the length of the question slice again I think, it's already done in the Resolve function